### PR TITLE
fix: Typescript config invalid with node v22.7.0 with ESM

### LINKED ISSUE #30084: Typescript config file is invalid with Node 22.7
### Current behavior

When using a .ts extension for your config file, start up fails. Changing it to. js fixes this.
This started happening to me from node 22.7, anything before that it works fine.

My closest and root package.json have both type: module.

`Your configFile is invalid: ./cypress.config.ts`

### Desired behavior

Compile and run tests

### Test code to reproduce

Have a config file with .ts extension and running node v22.7

Config file looks as this but shouldn't matter, it worked for 3 months or longer.

```js
import { defineConfig } from 'cypress';
import codeCoverageTask from '@cypress/code-coverage/task';

export default defineConfig({
    viewportHeight: 1000,
    viewportWidth: 1000,
    component: {
        devServer: {
            framework: 'react',
            bundler: 'vite',
        },
        setupNodeEvents(on, config) {
            codeCoverageTask(on, config);
            return config;
        },
    },
});
```

### Cypress Version

13.13.3

### Node version

v22.7.0

### Operating System

macOS 14.6.1

### Debug Logs

```shell
@boa/shared-ui:component-test:ci: Your configFile is invalid: /Users/xxxx/Sites/cit-frontend/packages/shared-ui/cypress.config.ts
@boa/shared-ui:component-test:ci:
@boa/shared-ui:component-test:ci: It threw an error when required, check the stack trace below:
@boa/shared-ui:component-test:ci:
@boa/shared-ui:component-test:ci: Error [ERR_INTERNAL_ASSERTION]: This is caused by either a bug in Node.js or incorrect usage of Node.js internals.
@boa/shared-ui:component-test:ci: Please open an issue with this stack trace at https://github.com/nodejs/node/issues
@boa/shared-ui:component-test:ci:
@boa/shared-ui:component-test:ci:     at assert (node:internal/assert:14:11)
@boa/shared-ui:component-test:ci:     at ModuleLoader.getModuleJobForRequire (node:internal/modules/esm/loader:329:5)
@boa/shared-ui:component-test:ci:     at new ModuleJobSync (node:internal/modules/esm/module_job:313:34)
@boa/shared-ui:component-test:ci:     at ModuleLoader.importSyncForRequire (node:internal/modules/esm/loader:314:11)
@boa/shared-ui:component-test:ci:     at loadESMFromCJS (node:internal/modules/cjs/loader:1381:24)
@boa/shared-ui:component-test:ci:     at Module._compile (node:internal/modules/cjs/loader:1503:5)
@boa/shared-ui:component-test:ci:     at Module._extensions..js (node:internal/modules/cjs/loader:1691:10)
@boa/shared-ui:component-test:ci:     at Module.load (node:internal/modules/cjs/loader:1317:32)
@boa/shared-ui:component-test:ci:     at Module._load (node:internal/modules/cjs/loader:1127:12)
@boa/shared-ui:component-test:ci:     at TracingChannel.traceSync (node:diagnostics_channel:315:14)
@boa/shared-ui:component-test:ci:     at wrapModuleLoad (node:internal/modules/cjs/loader:217:24)
@boa/shared-ui:component-test:ci:     at Module.require (node:internal/modules/cjs/loader:1339:12)
@boa/shared-ui:component-test:ci:     at require (node:internal/modules/helpers:126:16)
```

### Other

Node changelog: https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V22.md#2024-08-22-version-2270-current-rafaelgss

### PR DESCRIPTION
**Pull request by AtofStryker**
- Closes #30084

### Additional details

 Cypress with Typescript and ESM stopped working with node `22.7.0`. Based on the node [release notes](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V22.md?rgh-link-date=2024-08-22T15%3A58%3A27Z#2024-08-22-version-2270-current-rafaelgss) the `--experimental-detect-module` flag is now enabled by default which doesn't play nicely with the experimental ts_node module loader that cypress is using under the hood.

Turning it off via `--no-experimental-detect-module`  on the child process fixes the issue.

### Steps to test

Unit tests added to have some coverage in the `ProjectConfigIpc` and a binary system test  using node `22.7.0` has been added

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: cli/CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ _Released 8/27/2024 (PENDING)_
 **Bugfixes:**
 
 - Fixed an issue where files outside the Cypress project directory were not calculating the bundle output path correctly for the `file:preprocessor`. Addresses [#8599](https://github.com/cypress-io/cypress/issues/8599).
+- Fixed an issue where Cypress would not run if Nodejs version `22.7.0` was being used with Typescript and ES Modules. Fixes [#30084](https://github.com/cypress-io/cypress/issues/30084).
```
Comment (jennifer-shehane) [inline cli/CHANGELOG.md]:
```suggestion
- Fixed an issue where Cypress would not run if Node.js version `22.7.0` was being used with TypeScript and ES Modules. Fixes [#30084](https://github.com/cypress-io/cypress/issues/30084).
```

**Comment by AtofStryker** [comment]
System test fails without fix https://app.circleci.com/pipelines/github/cypress-io/cypress/63656/workflows/a843fd5a-1608-413f-baac-b5c191b8b02e/jobs/2638953

**Comment by MikeMcC399** [comment]
@AtofStryker 

I'm guessing that this issue might also occur if `experimental-detect-module` was explicitly enabled?

If so, then maybe mention this in the changelog.

It was introduced in [Node.js 20.10.0](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V20.md#20.10.0). See https://nodejs.org/docs/latest-v20.x/api/cli.html#--experimental-detect-module.

**Comment by AtofStryker** [comment]
@MikeMcC399 I gave this a try in[5ef8ef0](https://github.com/cypress-io/cypress/pull/30099/commits/5ef8ef0e4ceb44f5213c120ddba0fa82b717f1d0)

**Comment by MikeMcC399** [comment]
@AtofStryker 

- I opened https://github.com/cypress-io/cypress/issues/30101 separately to avoid overloading this PR.

**Comment by MikeMcC399** [comment]
@AtofStryker 

> I gave this a try in[5ef8ef0](https://github.com/cypress-io/cypress/pull/30099/commits/5ef8ef0e4ceb44f5213c120ddba0fa82b717f1d0)

- Looking at your code, it seems you only do something different with Node.js >= `22.7.0`, so the added comment is probably misplaced.
    > Cypress uses its own loader for TypeScript and ES Modules to parse the Cypress config and likely will not work with the `experimental-detect-module` Node.js option.

- I suggest reverting your addition to the changelog and dealing with https://github.com/cypress-io/cypress/issues/30101 for Node.js < `22.7.0`  separately. That would be cleaner I think.

File: packages/data-context/src/util/hasTypescript.ts [hunk 2]
```diff
@@ -1,5 +1,11 @@
 export function hasTypeScriptInstalled (projectRoot: string) {
   try {
+    // mocking this module is fairly difficult under unit test. We need to mock this for the ProjectConfigIpc unit tests
+    // as the scaffolded projects in the data-context package do not install dependencies related to the project.
+    if (process.env.CYPRESS_INTERNAL_MOCK_TYPESCRIPT_INSTALL === 'true') {
```
Comment (AtofStryker) [inline packages/data-context/src/util/hasTypescript.ts:5]: I couldn't figure out another way to test this. Not great, but at least gets us what we need

**Review by cacieprins — APPROVED** [review]

**Review by ryanthemanuel — APPROVED** [review]

**Pull request merged**